test/k8s/res_limits_test.go

```diff
@@ -0,0 +1,512 @@
+								LabelSelector: fmt.Sprintf("app=%s", Config.KumaServiceName),
+							})
+						Expect(err).ToNot(HaveOccurred())
+						pod := pods[0]
+						metrics, err := k8s.RunKubectlAndGetOutputE(
+							cluster.GetTesting(),
+							cluster.GetKubectlOptions(Config.KumaNamespace),
+							"exec", pod.Name, "-c", "control-plane", "--", "sh", "-c", "wget -O - http://localhost:5680/metrics")
+						Expect(err).ToNot(HaveOccurred())
+						printLog("control plane metrics: %s\n", metrics)
+					case e := <-watcher.ResultChan():
+						if e.Object == nil {
+							cancel()
+							return
+						}
+
+						pod, ok := e.Object.(*corev1.Pod)
+						if !ok {
+							continue
+						}
+
+						if e.Type == watch.Deleted || e.Type == watch.Added {
+							continue
+						}
+
+						if e.Type == watch.Error {
+							errCh <- errors.New("error watching pod")
+							continue
+						}
+
+						if pod.DeletionTimestamp != nil ||
+							pod.Status.Phase == "Pending" ||
+							k8s.IsPodAvailable(pod) {
+							continue
+						}
+
+						args := []string{"logs", "-c", "control-plane", pod.Name}
+						if hasPodContainerCrashed(pod) {
+							args = append(args, "-p")
+						}
+
+						podLogs, err := k8s.RunKubectlAndGetOutputE(
+							cluster.GetTesting(),
+							cluster.GetKubectlOptions(Config.KumaNamespace),
+							args...)
+						Expect(err).ToNot(HaveOccurred())
+
+						y, err := yaml.Marshal(pod)
+						if err != nil {
+							y = []byte(fmt.Sprintf("failed to marshal pod yaml: %v", err))
+						}
+						errCh <- fmt.Errorf("pod '%s' failed in phase '%s'; pod yaml: %s, pod logs: %s",
+							pod.Name, pod.Status.Phase, y, podLogs)
+						watcher.Stop()
+						return
+					case <-ctx2.Done():
+						watcher.Stop()
+						close(errCh)
+						return
+					}
+				}
+			}()
+		}
+
+		BeforeAll(func() {
+			cpuLimit := requireVar("PERF_LIMIT_MILLI_CPU")
+			i, err := strconv.Atoi(cpuLimit)
+			Expect(err).ToNot(HaveOccurred(), "invalid value of PERF_LIMIT_MILLI_CPU")
+			cpu = i
+
+			memLimit := requireVar("PERF_LIMIT_MEGA_MEMORY")
+			i, err = strconv.Atoi(memLimit)
+			Expect(err).ToNot(HaveOccurred(), "invalid value of PERF_LIMIT_MEGA_MEMORY")
+			memory = i
+
+			svcNum, _ := os.LookupEnv("PERF_TEST_NUM_SERVICES")
+			insNum, _ := os.LookupEnv("PERF_TEST_INSTANCES_PER_SERVICE")
+
+			if svcNum != "" {
+				numServices, _ = strconv.Atoi(svcNum)
+			}
+			if insNum != "" {
+				instancesPerService, _ = strconv.Atoi(insNum)
+			}
+		})
+
+		BeforeEach(func() {
+			Expect(cluster.Install(NamespaceWithSidecarInjection(TestNamespace))).To(Succeed())
+		})
+
+		AfterEach(func() {
+			Expect(cluster.DeleteNamespace(TestNamespace)).To(Succeed())
+		})
+
+		// should not crash, dump memory usage
+		It("should deploy all services and instances", func() {
+
+			adjustResource(cpu, memory, false)
+
+			deployDPs()
+
+			errCh := make(chan error)
+			ctx, cancelMonitoring := context.WithCancel(context.Background())
```

Comment:
We usually do `defer cancelMonitoring()` so we don't forget to call it on unexpected exists. For example what if `Expect(dpRet).To(BeTrue(), "data planes should be run and available")` fails?

Reply:
Sure, from coding perspective, it's a good practice to calling it using `defer`.